hack/gendiscoverycache/gendiscoverycache.go

```diff
@@ -0,0 +1,126 @@
+package main
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	configv1 "github.com/openshift/api/config/v1"
+	configclient "github.com/openshift/client-go/config/clientset/versioned"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/discovery/cached/disk"
+
+	"github.com/Azure/ARO-RP/pkg/database"
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/metrics/noop"
+	"github.com/Azure/ARO-RP/pkg/util/encryption"
+	utillog "github.com/Azure/ARO-RP/pkg/util/log"
+	"github.com/Azure/ARO-RP/pkg/util/restconfig"
+	"github.com/Azure/ARO-RP/pkg/util/version"
+)
+
+const discoveryCacheDir = "pkg/frontend/kubeactions/discovery/cache"
+
+func run(ctx context.Context, log *logrus.Entry) error {
+	if len(os.Args) != 2 {
+		return fmt.Errorf("usage: %s resourceid", os.Args[0])
+	}
+
+	_env, err := env.NewEnv(ctx, log)
+	if err != nil {
+		return err
+	}
+
+	cipher, err := encryption.NewXChaCha20Poly1305(ctx, _env, env.EncryptionSecretName)
+	if err != nil {
+		return err
+	}
+
+	db, err := database.NewDatabase(ctx, log.WithField("component", "database"), _env, &noop.Noop{}, cipher, "")
+	if err != nil {
+		return err
+	}
+
+	doc, err := db.OpenShiftClusters.Get(ctx, strings.ToLower(os.Args[1]))
+	if err != nil {
+		return err
+	}
+
+	restconfig, err := restconfig.RestConfig(_env, doc.OpenShiftCluster)
+	if err != nil {
+		return err
+	}
+
+	cvClient, err := configclient.NewForConfig(restconfig)
+	if err != nil {
+		return err
+	}
+
+	cli, err := disk.NewCachedDiscoveryClientForConfig(restconfig, discoveryCacheDir, "", 0)
```

Comment:
Not sure that I see how it makes code more bug proff, but I moved `RemoveAll` out of `writeAssets`. I still want to keep it as close to `writeAssets` as possible so we don't end up with removed cache if something errors before `writeAssets`. It's not critical: `git checkout -- pkg/frontend/kubeactions/discovery/cache/` and you have it back, so I'm happy to move it up if there is a reason why it's better to have it before `NewCachedDiscoveryClientForConfig`.

Also we don't need in `Mkdir`: `k8s.io/client-go/discovery/cached/disk.CachedDiscoveryClient` deals with creating directories for us (there are many of them, nested).

Reply:
The case I want to avoid is accidentally reusing a cache when we should have entirely invalidated it - that's why I want it above this line.  I'd rather see a removed cache on an error than accidentally failing to invalidate a cache.

Reply:
> The case I want to avoid is accidentally reusing a cache when we should have entirely invalidated it

@jim-minter that is the reason why we have `RemoveAll` at all: it's just a precautions, but it actually not needed because:

1. I pass TTL 0 into `disk.NewCachedDiscoveryClientForConfig(...)` which means that any existing cache will be treated as expired. But I wanted to have `RemoveAll` just in case `k8s.io/client-go/discovery/cached/disk` at some point will start handling 0 TTL differently.
1. Actual reading/writing of cache by `k8s.io/client-go/discovery/cached/disk` happens later: when we call `cli.ServerGroupsAndResources()` in `writeAssets` so having it above doesn't make any difference from the prespective of this point.

I moved `RemoveAll` up as you requestion as I don't have strong preferences. But it's not super necessary :)